raiden/transfer/mediated_transfer/target.py

```diff
@@ -39,12 +39,12 @@ def events_for_onchain_secretreveal(
         channel_state.reveal_timeout,
         block_number,
     )
-    secret_known = channel.is_secret_known(
+    secret_known_offchain = channel.is_secret_known_offchain(
         channel_state.partner_state,
         transfer.lock.secrethash,
     )
 
-    if not safe_to_wait and secret_known:
+    if not safe_to_wait and secret_known_offchain:
```

Comment:
Answering my own question... when the partner knows the secret via the SecretReveal message but the transfer hasn't been unlocked, we keep checking on every block until it becomes unsafe to wait then we invoke registering the secret on-chain.
- In this code path, the transfer hasn't been unlocked as the current target task still exists.
- Also in this code path, the secret is known (see: target.handle_block).
So if what the above doing is:
-  Check if partner knows the secret. If so, and it's not safe to wait, then register the secret on-chain.
What should happen is:
- Check if partner knows the secret, it's not safe and wait and that the secret hasn't been already registered on chain... if so, register the secret on-chain.

Reply:
> Check if partner knows the secret. If so, and it's not safe to wait, then register the secret on-chain. which on the next block, will happen again (please correct me if i am wrong here).

I believe there is a part of code that would prevent this.

At [events_for_onchain_secretreveal](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/secret_registry.py#L12) we got [this](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/secret_registry.py#L22) condition:

` if get_status(channel_state) in CHANNEL_STATES_PRIOR_TO_CLOSED + (CHANNEL_STATE_CLOSED, ):`

In here the `get_status(channel_state)` as you can see [here](https://github.com/raiden-network/raiden/blob/325f4327583dc8e75677751865e63dda600bd929/raiden/transfer/channel.py#L980) will return a `CHANNEL_STATE_SETTLED`, which would stop it from sending on on-chain transaction again.

This is because it remembers the last sent transaction and checks its status.

Reply:
So this still can't happen. I wrote a [test](https://github.com/raiden-network/raiden/pull/2846/commits/47af159e30d49052fb0215c5d2d9d3b54ec6bb10) for it.

The reason is that the [secret_known_offchain](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/mediated_transfer/target.py#L42) check fails [here](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/mediated_transfer/target.py#L42).

That is because when we [register the secret onchain](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/channel.py#L1574) we also [delete the lock](https://github.com/raiden-network/raiden/blob/47af159e30d49052fb0215c5d2d9d3b54ec6bb10/raiden/transfer/channel.py#L1549). This will remove it from the indexing structures and the next time around cause the `secret_known_offchain` check to fail.